File: app/src/main/java/com/ahao/myapplication/banner/BannerLayoutManager.java
package com.ahao.myapplication.banner;

import android.graphics.PointF;

import androidx.recyclerview.widget.OrientationHelper;
import androidx.recyclerview.widget.RecyclerView;

import android.view.View;

public class BannerLayoutManager extends RecyclerView.LayoutManager implements RecyclerView.SmoothScroller.ScrollVectorProvider {

    protected final OrientationHelper mOrientationHelper;

    private float heightScale = 0.9f;
    private float widthScale = 0.9f;

    private boolean loop = true;  //默认无限循环

    protected int itemWidth;

    private int smoothScrollTime = 500;

    private boolean hasLayout;

    public BannerLayoutManager() {
        mOrientationHelper = OrientationHelper.createHorizontalHelper(this);
    }

    @Override
    public RecyclerView.LayoutParams generateDefaultLayoutParams() {
        return new RecyclerView.LayoutParams(RecyclerView.LayoutParams.MATCH_PARENT, RecyclerView.LayoutParams.WRAP_CONTENT);
    }

    @Override
    public void onLayoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
        layoutChildren(recycler, state);
    }

    private void layoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
        if (getItemCount() == 0 || state.isPreLayout()) {
            removeAndRecycleAllViews(recycler);
            return;
        }

        if (hasLayout) {
            return;
        }

        detachAndScrapAttachedViews(recycler);

        View scrap = recycler.getViewForPosition(0);
        measureChildWithMargins(scrap, 0, 0);
        itemWidth = getDecoratedMeasuredWidth(scrap);
        int offsetX = (mOrientationHelper.getTotalSpace() - mOrientationHelper.getDecoratedMeasurement(scrap)) / 2;
        for (int i = 0; i < getItemCount(); i++) {
            if (offsetX > mOrientationHelper.getTotalSpace()) {
                break;
            }
            View viewForPosition = recycler.getViewForPosition(i);
            addView(viewForPosition);
            measureChildWithMargins(viewForPosition, 0, 0);
            offsetX += layoutItem(viewForPosition, offsetX);
        }


        View lastChild = getChildAt(getChildCount() - 1);
        // 如果是循环布局，并且最后一个view已超出父布局，则添加最左边的view
        if (loop && lastChild != null && getDecoratedRight(lastChild) > mOrientationHelper.getTotalSpace()) {
            layoutLeftItem(recycler);
        }
        doWithItem();
        hasLayout = true;
    }

    private void layoutLeftItem(RecyclerView.Recycler recycler) {
        View childCenter = getChildAt(getChildCount() - 2);
        if (childCenter != null) {
            View viewForPosition = recycler.getViewForPosition(getItemCount() - 1);
            addView(viewForPosition, 0);
            measureChildWithMargins(viewForPosition, 0, 0);
            int top = getItemTop(viewForPosition);
            int left = getDecoratedLeft(childCenter) - itemWidth;
            int right = left + itemWidth;
            layoutDecoratedWithMargins(viewForPosition, left, top, right, top + getDecoratedMeasuredHeight(viewForPosition));
        }
    }

    private int layoutItem(View viewForPosition, int offsetX) {
        layoutDecoratedWithMargins(viewForPosition, offsetX, getItemTop(viewForPosition), offsetX + itemWidth, getItemTop(viewForPosition) + getDecoratedMeasuredHeight(viewForPosition));
        return itemWidth;
    }

    @Override
    public boolean canScrollHorizontally() {
        return getItemCount() > 1;
    }

    @Override
    public int scrollHorizontallyBy(int dx, RecyclerView.Recycler recycler, RecyclerView.State state) {
        return recycler == null ? 0 : offsetDx(dx, recycler);
    }

    private int offsetDx(int dx, RecyclerView.Recycler recycler) {
        int realScroll = dx;
        // 向左
        if (dx > 0) {
            realScroll = scrollToLeft(dx, recycler);
        }
        // 向右
        if (dx < 0) {
            realScroll = scrollToRight(dx, recycler);
        }
        doWithItem();

        return realScroll;
    }

    protected void doWithItem() {
    }

    private int scrollToRight(int dx, RecyclerView.Recycler recycler) {
        int realScroll = dx;
        while (true) {
            View leftChild = getChildAt(0);
            int left = getDecoratedLeft(leftChild);
            if (left + Math.abs(dx) > getPaddingLeft()) {
                int position = getPosition(leftChild);
                if (!loop && position == 0) {
                    break;
                }

                int addPosition = loop ? (position - 1 + getItemCount()) % getItemCount() : position - 1;
                View addView = recycler.getViewForPosition(addPosition);
                addView(addView, 0);
                measureChildWithMargins(addView, 0, 0);
                layoutDecoratedWithMargins(addView, left - getDecoratedMeasuredWidth(addView), getItemTop(addView), left, getItemTop(addView) + getDecoratedMeasuredHeight(addView));

            } else {
                break;
            }
        }

        View firstChild = getChildAt(0);

        int right = getDecoratedRight(firstChild);
        if (getPosition(firstChild) == 0) {
            if (right + Math.abs(dx) > mOrientationHelper.getTotalSpace()) {
                realScroll = -(mOrientationHelper.getTotalSpace() - right);
            }
        }

        offsetChildrenHorizontal(-realScroll);

        for (int i = 0; i < getChildCount(); i++) {
            View childAt = getChildAt(i);
            int decoratedLeft = getDecoratedLeft(childAt);
            if (decoratedLeft > mOrientationHelper.getTotalSpace()) {
                removeAndRecycleView(childAt, recycler);
            }
        }
        return realScroll;
    }

    private int scrollToLeft(int dx, RecyclerView.Recycler recycler) {
        int realScroll = dx;
        while (true) {
            // 将需要添加的view添加到RecyclerView中
            View rightView = getChildAt(getChildCount() - 1);
            int decoratedRight = getDecoratedRight(rightView);
            if (decoratedRight - dx < mOrientationHelper.getTotalSpace()) {
                int position = getPosition(rightView);
                if (!loop && position == getItemCount() - 1) {
                    break;
                }

                int addPosition = loop ? (position + 1) % getItemCount() : position + 1;
                View lastViewAdd = recycler.getViewForPosition(addPosition);
                addView(lastViewAdd);
                measureChildWithMargins(lastViewAdd, 0, 0);
                int left = decoratedRight;
                layoutDecoratedWithMargins(lastViewAdd, left, getItemTop(lastViewAdd), left + getDecoratedMeasuredWidth(lastViewAdd), getItemTop(lastViewAdd) + getDecoratedMeasuredHeight(lastViewAdd));
            } else {
                break;
            }
        }

        // 处理滑动
        View lastChild = getChildAt(getChildCount() - 1);
        int left = getDecoratedLeft(lastChild);
        if (getPosition(lastChild) == getItemCount() - 1) {
            // 最后一个view已经到底了，计算实际可以滑动的距离
            if (left - dx < 0) {
                realScroll = left;
            }
        }
        offsetChildrenHorizontal(-realScroll);

        // 回收滑出父布局的view
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            int decoratedRight = getDecoratedRight(child);
            if (decoratedRight < 0) {
                removeAndRecycleView(child, recycler);
            }
        }
        return realScroll;
    }

    public int getCurrentPosition() {
        for (int i = 0; i < getChildCount(); i++) {
            View childAt = getChildAt(i);
            if (getDecoratedLeft(childAt) >= 0 && getDecoratedRight(childAt) <= mOrientationHelper.getTotalSpace()) {
                return getPosition(childAt);
            }
        }
        return -1;
    }

    @Override
    public void smoothScrollToPosition(RecyclerView recyclerView, RecyclerView.State state, int targetPosition) {
        if (!loop && (targetPosition < 0 || targetPosition > getItemCount() - 1)) {
            return;
        }
        if (loop || getItemCount() > 0) {
            targetPosition = (targetPosition % getItemCount() + getItemCount()) % getItemCount();
        }

        int offset;
        recyclerView.requestFocus();
        int currentPosition = getCurrentPosition();
        if (currentPosition == getItemCount() - 1 && targetPosition == 0 && loop) {
            offset = itemWidth;
        } else {
            offset = (targetPosition - currentPosition) * itemWidth;
        }
        recyclerView.smoothScrollBy(offset, 0, null, smoothScrollTime);
    }

    private int getTotalHeight() {
        return getHeight() - getPaddingTop() - getPaddingBottom();
    }

    private int getItemTop(View item) {
        return (getTotalHeight() - getDecoratedMeasuredHeight(item)) / 2 + getPaddingTop();
    }

    public boolean isLoop() {
        return loop;
    }

    public void setLoop(boolean loop) {
        this.loop = loop;
    }


    @Override
    public PointF computeScrollVectorForPosition(int targetPosition) {
        return null;
    }

    public int getSmoothScrollTime() {
        return smoothScrollTime;
    }

    public void setSmoothScrollTime(int smoothScrollTime) {
        this.smoothScrollTime = smoothScrollTime;
    }

    public float getHeightScale() {
        return heightScale;
    }

    public void setHeightScale(float heightScale) {
        this.heightScale = heightScale;
    }

    public float getWidthScale() {
        return widthScale;
    }

    public void setWidthScale(float widthScale) {
        this.widthScale = widthScale;
    }
}

File: bannerview/src/main/java/com/ahao/bannerview/BannerLayoutManager.kt
package com.ahao.bannerview

import android.graphics.PointF
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.RecyclerView
import android.view.View
import kotlin.math.abs

open class BannerLayoutManager : RecyclerView.LayoutManager(), RecyclerView.SmoothScroller.ScrollVectorProvider {

    var mOrientationHelper: OrientationHelper = OrientationHelper.createHorizontalHelper(this)
    var itemWidth: Int = 0

    private var hasLayout = false
    var smoothScrollTime = 500
    var loop: Boolean = true

    override fun generateDefaultLayoutParams(): RecyclerView.LayoutParams {
        return RecyclerView.LayoutParams(RecyclerView.LayoutParams.MATCH_PARENT, RecyclerView.LayoutParams.WRAP_CONTENT)
    }

    override fun onLayoutChildren(recycler: RecyclerView.Recycler?, state: RecyclerView.State?) {
        recycler!!
        state!!
        layoutChildren(recycler, state)
    }

    private fun layoutChildren(recycler: RecyclerView.Recycler, state: RecyclerView.State) {
        if (itemCount == 0 || state.isPreLayout) {
            removeAndRecycleAllViews(recycler)
            return
        }
        if (hasLayout) {
            return
        }
        detachAndScrapAttachedViews(recycler)

        val scrap = recycler.getViewForPosition(0)
        measureChildWithMargins(scrap, 0, 0)
        itemWidth = getDecoratedMeasuredWidth(scrap)
        var offsetX = (mOrientationHelper.totalSpace - mOrientationHelper.getDecoratedMeasurement(scrap)) / 2

        for (i in 0 until itemCount) {
            if (offsetX > mOrientationHelper.totalSpace) {
                break
            }
            val viewForPosition = recycler.getViewForPosition(i)
            addView(viewForPosition)
            measureChildWithMargins(viewForPosition, 0, 0)
            offsetX += layoutItem(viewForPosition, offsetX)
        }
        if (itemCount >= 3 && loop) {
            layoutLeftItem(recycler)
        }
        doWithItem()

        hasLayout = true
    }

    private fun layoutItem(viewForPosition: View, offsetX: Int): Int {
        layoutDecoratedWithMargins(viewForPosition, offsetX, getItemTop(viewForPosition), offsetX + itemWidth, getItemTop(viewForPosition) + getDecoratedMeasuredHeight(viewForPosition))
        return itemWidth
    }

    private fun layoutLeftItem(recycler: RecyclerView.Recycler) {
        val childCenter = getChildAt(childCount - 2)
        if (childCenter != null) {
            val viewForPosition = recycler.getViewForPosition(itemCount - 1)
            addView(viewForPosition, 0)
            measureChildWithMargins(viewForPosition, 0, 0)
            val top = getItemTop(viewForPosition)
            val left = getDecoratedLeft(childCenter) - itemWidth
            val right = left + itemWidth
            layoutDecoratedWithMargins(viewForPosition, left, top, right, top + getDecoratedMeasuredHeight(viewForPosition))
        }
    }

    override fun canScrollHorizontally(): Boolean {
        return itemCount > 1
    }

    override fun scrollHorizontallyBy(dx: Int, recycler: RecyclerView.Recycler?, state: RecyclerView.State?): Int {
        return if (recycler == null) 0 else offsetDx(dx, recycler)
    }

    private fun offsetDx(dx: Int, recycler: RecyclerView.Recycler): Int {
        var realScroll = dx
        // 向左
        if (dx > 0) {
            realScroll = scrollToLeft(dx, recycler)
        }
        // 向右
        if (dx < 0) {
            realScroll = scrollToRight(dx, recycler)
        }
        doWithItem()

        return realScroll
    }

    open fun doWithItem() {}

    private fun scrollToRight(dx: Int, recycler: RecyclerView.Recycler): Int {
        var realScroll = dx
        while (true) {
            val leftChild = getChildAt(0)
            val left = getDecoratedLeft(leftChild!!)
            if (left + abs(dx) > paddingLeft) {
                val position = getPosition(leftChild)
                if (!loop && position == 0) {
                    break
                }

                val addPosition = if (loop) (position - 1 + itemCount) % itemCount else position - 1
                val addView = recycler.getViewForPosition(addPosition)
                addView(addView, 0)
                measureChildWithMargins(addView, 0, 0)
                layoutDecoratedWithMargins(addView, left - getDecoratedMeasuredWidth(addView), getItemTop(addView), left, getItemTop(addView) + getDecoratedMeasuredHeight(addView))
            } else {
                break
            }
        }

        val firstChild = getChildAt(0)
        val right = getDecoratedRight(firstChild!!)
        if (getPosition(firstChild) == 0) {
            if (right + abs(dx) > mOrientationHelper.totalSpace) {
                realScroll = -(mOrientationHelper.totalSpace - right)
            }
        }

        offsetChildrenHorizontal(-realScroll)

        (0..childCount).mapNotNull {
            getChildAt(it)
        }.forEach {
            if (getDecoratedLeft(it) > mOrientationHelper.totalSpace) {
                removeAndRecycleView(it, recycler)
            }
        }
        return realScroll
    }

    private fun scrollToLeft(dx: Int, recycler: RecyclerView.Recycler): Int {
        var realScroll = dx
        while (true) {
            val rightView = getChildAt(childCount - 1)
            val decoratedRight = getDecoratedRight(rightView!!)
            if (decoratedRight - dx < mOrientationHelper.totalSpace) {
                val position = getPosition(rightView)
                if (!loop && position == itemCount - 1) {
                    break
                }

                val addPosition = if (loop) (position + 1) % itemCount else position + 1
                val lastViewAdd = recycler.getViewForPosition(addPosition)
                addView(lastViewAdd)
                measureChildWithMargins(lastViewAdd, 0, 0)
                layoutDecoratedWithMargins(lastViewAdd, decoratedRight, getItemTop(lastViewAdd), decoratedRight + getDecoratedMeasuredWidth(lastViewAdd), getItemTop(lastViewAdd) + getDecoratedMeasuredHeight(lastViewAdd))
            } else {
                break
            }
        }

        val lastChild = getChildAt(childCount - 1)
        val left = getDecoratedLeft(lastChild!!)
        if (getPosition(lastChild) == itemCount - 1) {
            if (left - dx < 0) {
                realScroll = left
            }
        }
        offsetChildrenHorizontal(-realScroll)

        (0..childCount).mapNotNull {
            getChildAt(it)
        }.forEach {
            val decoratedRight = getDecoratedRight(it)
            if (decoratedRight < 0) {
                removeAndRecycleView(it, recycler)
            }
        }
        return realScroll
    }

    fun getCurrentPosition(): Int {
        for (i in 0 until childCount) {
            val childAt = getChildAt(i) ?: continue
            if (getDecoratedLeft(childAt) >= 0 && getDecoratedRight(childAt) <= mOrientationHelper.totalSpace) {
                return getPosition(childAt)
            }
        }
        return -1
    }

    override fun smoothScrollToPosition(recyclerView: RecyclerView?, state: RecyclerView.State?, position: Int) {
        var targetPosition = position
        if (!loop && (targetPosition < 0 || targetPosition > itemCount - 1)) {
            return
        }
        if (loop || itemCount > 0) {
            targetPosition = (targetPosition % itemCount + itemCount) % itemCount
        }

        recyclerView!!.requestFocus()
        val currentPosition = getCurrentPosition()
        val offset = if (currentPosition == itemCount - 1 && targetPosition == 0 && loop) {
            itemWidth
        } else {
            (targetPosition - currentPosition) * itemWidth
        }

        recyclerView.smoothScrollBy(offset, 0, null, smoothScrollTime)

    }

    private fun getTotalHeight(): Int {
        return height - paddingTop - paddingBottom
    }

    private fun getItemTop(item: View): Int {
        return (getTotalHeight() - getDecoratedMeasuredHeight(item)) / 2 + paddingTop
    }

    override fun computeScrollVectorForPosition(target: Int): PointF? = null
}